### PR TITLE
Add test for Test module.

### PR DESCRIPTION
**Pull request by kostiakoval**
Check that Package.xctest are generated.

### CODE REVIEW

File: Tests/Functional/TestMiscellaneous.swift [hunk 1]
```diff
@@ -303,4 +303,13 @@ class MiscellaneousTestCase: XCTestCase {
             XCTAssertEqual(output, "Goodbye\n")
         }
     }
+
+    func testModuleWithTests() {
+        fixture(name: "SwiftTesting/SingleTarget", file: #file, line: #line) { prefix in
+            XCTAssertBuilds(prefix)
+            print(prefix)
```
Comment (aciidb0mb3r) [inline Tests/Functional/TestMiscellaneous.swift]: should print be there in tests?

Comment (kostiakoval) [inline Tests/Functional/TestMiscellaneous.swift]: :+1: Thanks. Fixed

**Comment by modocache** [comment]
Awesome! Would love to see this and more tests for the test module merged.

**Comment by kostiakoval** [comment]
Me to :) Not sure if this test, and other, related to tests module, should be in a separate file?

**Comment by mxcl** [comment]
On Linux it has a different name. Also needs the other Linux machinery.

**Comment by kostiakoval** [comment]
Like that? I haven't tested it on linux, hope it will work :)

**Comment by mxcl** [comment]
Not quite, there is only one `LinuxMain.swift`.

**Comment by kostiakoval** [comment]
`LinuxMain.swift` is for a Fixture, so I can run test manually on the Linux.
I've update it to latest testing syntax and tested on Linux, it works.

Is there anything I should with it?

**Comment by aciidb0mb3r** [comment]
👌

File: Tests/Functional/TestMiscellaneous.swift [hunk 2]
```diff
@@ -303,4 +303,18 @@ class MiscellaneousTestCase: XCTestCase {
...
+
+    func testModuleWithTests() {
+        fixture(name: "SwiftTesting/SingleTarget", file: #file, line: #line) { prefix in
+            XCTAssertBuilds(prefix)
+            XCTAssertFileExists(prefix, ".build", "debug", "Foo.swiftmodule")
+            #if os(OSX)
+                XCTAssertDirectoryExists(prefix, ".build", "debug", "Package.xctest")
+            #else
+                //FIXME: test-Package not generated during swift-build on linux so this will fail.
+                //XCTAssertFileExists(prefix, ".build", "debug", "test-Package")
```
Comment (mxcl) [inline Tests/Functional/TestMiscellaneous.swift]: This should work, I'm confused, something doesn't work?

Comment (aciidb0mb3r) [inline Tests/Functional/TestMiscellaneous.swift]: `XCTAssertBuilds` calls `executeSwiftBuild` which runs `swift build`
`test-Package` executable is generated when "test" target is built

Comment (mxcl) [inline Tests/Functional/TestMiscellaneous.swift]: Of course. Well, we should run `swift test`.

Comment (kostiakoval) [inline Tests/Functional/TestMiscellaneous.swift]: Ok, I will fix it. Thanks for feedback

**Comment by kostiakoval** [comment]
@mxcl I've added `XCTAssertTests` to run `swift-test`. Works both on Mac and Linux.

**Comment by mxcl** [comment]
@swift-ci Please test

**Comment by kostiakoval** [comment]
@modocache Can I get any help why tests are failing? 
```
tmp/../SingleTarget/Tests/Foo/Test.swift:2:8: error: no such module 'XCTest'
import XCTest
```

**Comment by mxcl** [comment]
@kostiakoval platform? On OS X don't build swift with xctest, on Linux you also have to build foundation now and depending on how you invoke the tests, tell the system where to find it.

**Comment by kostiakoval** [comment]
I'm not sure how to pass those parameters. From what I see we can't pass any arguments to `swift-test` on linux. Should I send some environment variables? 

I would appreciate any help please .

**Comment by modocache** [comment]
Thanks again for working on this, @kostiakoval! I think it'll be really cool to not only test corelibs-xctest's behavior (which we do in corelibs-xctest), but also the behavior of the `swift test` command.

First off as a basic sanity check I pulled down the latest revision on master for all repositories onto my Ubuntu 15.10 machine and ran:

```
$ utils/build-script -R --swiftpm --xctest --test -- \
    --skip-test-cmark --skip-test-swift --skip-test-foundation
```

This succeeded. I then tried to pull in the changes from your pull request locally:

```
$ cd ../swiftpm
$ git pull https://github.com/kostiakoval/swift-package-manager.git TestTarget-tests
```

Re-running the Swift `utils/build-script` from before, I get the following build error:

```
Compiling Swift Module 'Gettest' (4 sources)
/home/modocache/GitHub/apple/swiftpm/Tests/Get/Utilities.swift:136:20: error: use of unresolved identifier 'Resources'
    let toolPath = Resources.findExecutable("swift-test")
                   ^~~~~~~~~
/home/modocache/GitHub/apple/swiftpm/Tests/Transmute/Utilities.swift:136:20: error: use of unresolved identifier 'Resources'
    let toolPath = Resources.findExecutable("swift-test")
                   ^~~~~~~~~
/home/modocache/GitHub/apple/swiftpm/Tests/Functional/Utilities.swift:136:20: error: use of unresolved identifier 'Resources'
    let toolPath = Resources.findExecutable("swift-test")
                   ^~~~~~~~~
/home/modocache/GitHub/apple/swiftpm/Tests/ManifestParser/Utilities.swift:136:20: error: use of unresolved identifier 'Resources'
    let toolPath = Resources.findExecutable("swift-test")
                   ^~~~~~~~~
<unknown>:0: error: build had 4 command failures
error: exit(1): /home/modocache/GitHub/apple/build/Ninja-ReleaseAssert/llbuild-linux-x86_64/bin/swift-build-tool -f /home/modocache/GitHub/apple/build/Ninja-ReleaseAssert/swiftpm-linux-x86_64/debug.yaml test
bootstrap: error: tests failed with exit status 1
utils/build-script: command terminated with a non-zero exit status 1, aborting
```

I seem to recall that `Resources` was renamed or removed in a recent commit (I think it may have been Xcodeproj support), so your commit probably needs to be updated to reflect that.

Anyway, silly me, I guess I probably should have tested exactly your changes, and not them merged into `master`! :sweat_smile: I tried just that:

```
$ git remote add kostiakoval https://github.com/kostiakoval/swift-package-manager.git
$ git fetch kostiakoval 
$ git checkout TestTarget-tests
```

But unfortunately this time I get Swift 3-related errors.

I'll try to checkout pre-Swift 3 versions of all the Swift libraries and re-run your pull request. In the meantime, if you could rebase your changes onto the latest swiftpm master branch and fix up the build errors, that'd be super helpful! :bow:

**Comment by modocache** [comment]
Because of the above I'm not able to test this out myself, but my current theory is that you'll need to pass flags to `executeSwiftTest()`, as the test utilities in SwiftPM already do for `executeSwiftBuild()` and `XCTAssertBuilds()`.

In the `ModuleMapsTestCase`, for example, I see that `-L /some/output/directory` is passed to `XCTAssertBuilds()`. You'll probably need to do the same with `executeSwiftTest()`, but passing some subset of [the flags I pass here](https://github.com/apple/swift-package-manager/commit/1cabbde47ce10d97e8328d49bd44720c7b58b695#diff-296aa080a049a48a75c47e0a2926fde0R553).

A corelibs-foundation framework is included with Swift snapshot releases. If you're interested in having the user use the currently installed version when testing, then you can link that one. But I imagine that in order for this test to be robust, you'd want to allow a path to a Foundation framework to be specified somehow...

**Comment by kostiakoval** [comment]
@modocache Thanks for you amazing feedback! 
I've fixed the `Resources.findExecutable("swift-test")` build error. Tests are working again on MacOS. 

@mxcl Can we run CI to see if we still have that error on Linux?

**Comment by mxcl** [comment]
@swift-ci Please test

**Comment by kostiakoval** [comment]
When I try to run test on Linux, even from the master repo, I get this error

./Utilities/bootstrap test
```
/home/kkoval/apple/swiftpm/Tests/PackageType/PackageNameTests.swift:12:8: error: no such module 'XCTest'
import XCTest
...
```

**Comment by mxcl** [comment]
On linux you have to specify the path to: `--xctest`

See `Utilities/bootstrap --help`.

**Comment by modocache** [comment]
> On linux you have to specify the path to: `--xctest`

You now must also specify the path to Foundation, using `--foundation`. I've been meaning to add this to the README.

**Comment by mxcl** [comment]
Honestly it is better to run the tests via the swift build-script. I use:

    swift/utils/build-script \
        --swiftpm --llbuild --xctest --foundation --test --release -- \
        --skip-test-swift --skip-test-cmark --skip-test-xctest --skip-test-foundation

**Comment by kostiakoval** [comment]
Yes that's right. But how can I specify `--xctest` path when I call `swift-test`.
Right now when we call `swift-test` on Linux we don't pass any arguments to swift-corelibs-xctest. 
https://github.com/apple/swift-package-manager/blob/master/Sources/swift-test/test.swift#L26

Does swift-corelibs-xctest support it now?

**Comment by aciidb0mb3r** [comment]
@kostiakoval I was able to fix by just changing this https://github.com/aciidb0mb3r/swift-package-manager/commit/772335dbbe8f827f790cced7c7c248d71de489d3 
I built toolchain on my Linux VM with 
```
$ ./swift/utils/build-script --preset="buildbot_linux" install_destdir=/home/aciid/hdd/toolchain/swift-nightly installable_package=/home/aciid/hdd/toolchain/swift.tar.gz
```

**Comment by mxcl** [comment]
>  But how can I specify --xctest path when I call swift-test.

You can't, this is not a supported execution model. `swift-test` must be part of a complete toolchain or it must use a complete toolchain via the `TOOLCHAINS` environment variable.

**Comment by kostiakoval** [comment]
The problem was in wrong naming. I've tested it local and it works. Can we invoke CI and get it merged :)

**Comment by mxcl** [comment]
Conflicts.

**Comment by kostiakoval** [comment]
@mxcl  Fixed. And cleaned up commits

**Comment by kostiakoval** [comment]
@mxcl Could you please invoke "CI test" and merge it please? Also PR #122  is ready

**Comment by aciidb0mb3r** [comment]
👍

**Comment by mxcl** [comment]
@swift-ci Please test

**Comment by czechboy0** [comment]
@kostiakoval Tests failed on Linux, can you please take a :eyes: ?

**Comment by kostiakoval** [comment]
I found how to reproduce that bug. It works fine if you run `./Utils/bootstrap test` with no extra parameters. 
I happens when you run it and provide `--swiftc --xctext --foundation --sbt` options. 
Now I need to find a way to fix it :)

**Comment by mxcl** [comment]
I've tried to make this work on Linux and my conclusion is we shouldn't bother with this test:

1. It is hard to make this work in the context of a full build of swift since the paths for XCTest and Foundation are not in the standard "toolchain" shape.
2. We are always testing swift-test because we use swift-test to test ourselves.

Instead we could add an integration test, but really we don't even need to because we test swift-test when we test ourselves.

So I'd like to close this, if there are no objections.

**Comment by kostiakoval** [comment]
I agree with point 1. I've opened that RP because some `Tests` layout didn't work
https://github.com/apple/swift-package-manager/pull/162

I agree to close that PR and find a better way to test it.

**Comment by czechboy0** [comment]
@mxcl If we can find a way to still implement something like `XCTAssertBuilds` but for tests, I'm fine with closing this one. But I definitely think we need to validate layouts not just with `swift-build`, but also with `swift-test`, because completely valid layouts have (twice already) been broken on a released snapshot. I believe we could have prevented the second breakage if we had `XCTAssertTests`, because in that case `swift-build` worked but `swift-test` failed to compile tests. So I'm only OK with closing this PR if there's a strategy you have in mind that can give us the ability to validate a layout for `swift-test`. I'm afraid that if we don't start testing out layouts completely, sooner or later we'll break `swift-test` again for projects in the wild.

@kostiakoval Exactly why I'm following this PR - I had my projects break on a snapshot bc `swift-test` didn't compile test on a layout that was only being tested by running `swift-build`.

**Comment by mxcl** [comment]
This PR doesn't test layouts. But I agree we should test layouts. We don't have to do it functionally. Would be better to have unit tests for that component.

**Pull request closed without merging**